--- a/tools/gn/tools/gn/args.cc
+++ b/tools/gn/tools/gn/args.cc
@@ -320,9 +320,10 @@ void Args::SetSystemVarsLocked(Scope* dest) const {
     arch = kX86;
   else if (os_arch == "x86_64")
     arch = kX64;
-  else if (os_arch.substr(0, 3) == "arm")
+  else if (os_arch.substr(0, 3) == "arm" && os_arch[4] < '8')
     arch = kArm;
-  else if (os_arch == "aarch64")
+  else if ((os_arch == "aarch64") ||
+           (os_arch.substr(0, 3) == "arm" && os_arch[4] >= '8'))
     arch = kArm64;
   else if (os_arch == "mips")
     arch = kMips;
